src/scanner.rs

```diff
@@ -1,63 +1,100 @@
+    }
+    pub fn print_tokens(&self) {
+        println!("tokens: {:?}", self.tokens);
+    }
+    fn scan_token(&mut self, rox: &mut Rox) {
+        match self.advance() {
+            '(' => self.add_token(TokenType::LeftParen),
+            ')' => self.add_token(TokenType::RightParen),
+            '{' => self.add_token(TokenType::LeftBrace),
+            '}' => self.add_token(TokenType::RightBrace),
+            ',' => self.add_token(TokenType::Comma),
+            '.' => self.add_token(TokenType::Dot),
+            '-' => self.add_token(TokenType::Minus),
+            '+' => self.add_token(TokenType::Plus),
+            ';' => self.add_token(TokenType::Semicolon),
+            '*' => self.add_token(TokenType::Star),
+            '!' => if self._match('=') {
+                self.add_token(TokenType::BangEqual)
+            } else {
+                self.add_token(TokenType::Bang)
+            },
+            '=' => if self._match('=') {
+                self.add_token(TokenType::EqualEqual)
             } else {
-                Scanner::run(&line.trim().to_string(), state);
-                line.clear();
-                state.had_error = false;
-            }
+                self.add_token(TokenType::Equal)
+            },
+            '<' => if self._match('=') {
+                self.add_token(TokenType::LessEqual)
+            } else {
+                self.add_token(TokenType::Less)
+            },
+            '>' => if self._match('=') {
+                self.add_token(TokenType::GreaterEqual)
+            } else {
+                self.add_token(TokenType::Greater)
+            },
+            c @ _ => Rox::error(self.line, format!("Unrecognized character: {}", c), rox),
         }
-        println!("done");
     }
-    fn run(source: &String, state: &mut GlobalState) {
-        let scanner: Scanner = Scanner { source: source.to_string() };
-        let tokens: Vec<Token> = scanner.scan_tokens();
-        for token in tokens {
-            println!("{:?}", token._type);
+    fn add_token(&mut self, _type: TokenType) {
+        self.tokens.push(Token {
+            _type: _type,
+            lexeme: None,
+            literal: None,
+            line: self.line,
+        });
+    }
+    fn advance(&mut self) -> char {
+        self.current = self.current + 1;
+        self.get_source_char(self.current - 1)
+    }
+    fn _match(&mut self, c: char) -> bool {
+        if self.is_at_end() {
+            return false;
+        }
+        if self.get_source_chars()[self.current] != c {
+            return false;
         }
-        Scanner::error(2, "I can't believe it's not butter.".to_string(), state);
+        self.current = self.current + 1;
+        true
     }
-    fn error(line: u32, message: String, state: &mut GlobalState) {
-        Scanner::report(line, "".to_string(), message, state);
+    fn get_source_char(&self, index: usize) -> char {
+        self.get_source_chars()[index]
```

Comment:
Not sure if it's a good idea to capture out of bounds exceptions here by using the `try!()` macro.